flume-ng-core/src/main/java/org/apache/flume/source/ExecSource.java

```diff
@@ -333,7 +333,7 @@ public void run() {
           while ((line = reader.readLine()) != null) {
             sourceCounter.incrementEventReceivedCount();
             synchronized (eventList) {
-              eventList.add(EventBuilder.withBody(line.getBytes(charset)));
+              eventList.add(EventBuilder.withBody(line.getBytes("UTF-8")));
```

Comment:
The `charset` variable you just removed contains the configurable charset which is `UTF-8` by default if you have not changed it in your configuration.

Did this change actually fix your issue?

Reply:
Yes, it is solved.
Look roughly at the data flow:
If I specify ‘a1.sources.r1.charset = GBK’ in the source configuration, the whole process of the data is encoded as follows:
ExecSource read data (GBK) -->> Data write eventList use encoding (GBK) --->> RegexExtractorInterceptor.java line 130 decoding (UTF-8, garbled) --->> elasticsearchSink default decoding (UTF-8 and no configuration items, Garbled)
So, I think that after reading out the data according to the configuration code, it should use the default UTF-8 encoding (getBytes()) instead of the configured encoding (GBK), otherwise the subsequent decoding (UTF- 8) Will be garbled.
----------------------------------------------------------------------原文，我的英语太糟糕
是的，解决了。
大致看下数据流程：
如果我在source配置中指定‘a1.sources.r1.charset = GBK’，那么数据整个过程编码如下：
execSource读数据（GBK）>> 数据写入eventList使用编码（GBK） >> RegexExtractorInterceptor.java第130行解码用（UTF-8，已乱码）>> elasticsearchSink默认解码（UTF-8且没有配置项，已乱码）
所以，我认为应该按照配置编码读取出数据后，应该使用使用默认的UTF-8来编码（getBytes（）），而不应该按照配置的编码（GBK）来编码，否则后续的解码（UTF-8）会乱码。

Reply:
I see. Thanks for the clarification. In this case, the charset change should be either in the elasticsearchsink or should be a new config parameter like `targerCharset`